pkg/controller/node/nodecontroller.go

```diff
@@ -910,3 +1033,77 @@ func (nc *NodeController) ComputeZoneState(nodeReadyConditions []*api.NodeCondit
 		return notReadyNodes, stateNormal
 	}
 }
+
+func tryAddTaintToNode(kubeClient clientset.Interface, nodeName string, taintToAdd api.Taint) (bool, error) {
+	node, err := kubeClient.Core().Nodes().Get(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	oldTaints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		return false, err
+	}
+
+	for _, oldTaint := range oldTaints {
+		if taintToAdd.MatchTaint(oldTaint) {
+			return false, nil
```

Comment:
If taint already exists, this function will do nothing, and returning false here means taint is not added this time.